zerver/lib/url_decoding.py

```diff
@@ -66,3 +66,56 @@ def is_same_server_message_link(url: str) -> bool:
     ends_with_near_message_id = fragment_parts[-2] == "near" and fragment_parts[-1].isdigit()
 
     return ends_with_near_message_id
+
+
+class NearLinkHandler:
+    """
+    The NearLinkHandler is a helper class for editing
+    and cleaning up near links.
+
+    It can do basic operations such as splitting, fetching
+    link parts, and reassembling. It also applies some near
+    link related validations and clean-up operations.
+
+    See `test_near_link_variations.json` for examples of
+    links this class is intended to handle.
+    """
+
+    def __init__(self, near_link: str) -> None:
+        if not check_near_link_base(near_link):
+            raise AssertionError("This near link is either invalid or not from this server.")
+        self.split_result: SplitResult
+        self.patch_near_link(urlsplit(near_link))
+
+    def clean_near_link(self, split_result: SplitResult) -> SplitResult:
+        """
+        This function fixes legacy near links (uses "stream"),
+        and makes sure relative links starts with "/".
+        """
+        fragment_parts = split_result.fragment.split("/")
+        changed_parts = {}
+
+        if fragment_parts[1] == "stream":
+            fragment_parts[1] = "channel"
+        if split_result.hostname is None and split_result.path == "":
+            # Makes sure a relative near link starts with "/"
+            changed_parts["path"] = "/"
+
+        fragments = "/".join(fragment_parts)
+        changed_parts["fragment"] = fragments
+        cleaned_split_result = split_result._replace(**changed_parts)
+        return cleaned_split_result
```

Comment:
I think it's particularly visible here, but we have the wrong data structures for here, and it results in the parsing being wrong in corner cases that no known code produces, but should be valid. 

I think what you want to be doing is parsing the URL into a `list[NarrowParameter]` structure, and then look up the `channel` term using functions like this one in `filter.ts`.

```
    operands(operator: string): string[] {                                                              
        return this._terms                                                                              
            .filter((term) => !term.negated && term.operator === operator)                              
            .map((term) => term.operand);                                                               
    }                                                                                                   
```

It might make sense to just do a prep PR doing the necessary parsing infrastructure, and then rework this to use that. It could be:

- First commit: Rename `NarrowTerm` in `zerver/lib/narrow_helpers.py` to `NeverNegatedNarrowTerm`, which can be a subclass of a `NarrowTerm` class that has the missing `negated` flag.
- Add a  `parse_narrow_url` in `url_decoding.py` that parses an arbitrary narrow URL into a `List[NarrowTerm]`, broadly following the approach `hash_util.parse_narrow`, with tests.

And then we can merge that, and this can then be implemented on top of that helper -- and would probably involve writing something a bit like a super-simplified form of `class Filter` in the web app to support these cleaning operations but with nice abstractions for "get me the channel operator, if any", but without trying to access any data (like checking if the stream ID is valid, etc.). The overall goal here would be to have as little of the more complex logic as possible interact with a URL with all the messy `/` splitting as possible -- we'll just generate a `List[NarrowTerm]`, use it to compute a new one that's modernized and has updated IDs, and then turn that `List[NarrowTerm]` into a new URL.

(A reason this is important, beyond just safety/correctness if someone writes a `/#narrow/topic/10/channel/11/near/16` URL, which by the spec is supposed to work just like if the operators weren't in a funny order, is that we're going to need to extend what this PR does with `near` links to support `with/{message_id}` too, and at some point you just want an actual parser into nice data structures)